File: Java/14-Ordenação de Vetores/ShellSort.java
class ShellSort {
    // Podemos ordenar um vetor através de alguns métodos criados por programadores
    // Este método é chamado ShellSort
    static int[] shellSort(int[] vetor){
        int h = 1;
        int n = vetor.length;
        
        while(h < n) {
            h *= 3 + 1;
        }
        
        h /= 3;
        int c;
        int j;
        
        while(h > 0) {
            for(int i = h; i < n; i++) {
                c = vetor[i];
                j = i;
                while(j >= h && vetor[j - h] > c) {
                    
                    vetor[j] = vetor[j - h];
                    j -= h;
                }
                
                vetor[j] = c;
            }
            
            h /= 2;
        }
        
        return vetor;
    }

    public static void main(String[] args) {
        // Vetor esta fora de ordem
        int[] vet = { 4, 2, 1, 8, 3 };

        System.out.println("Vetor original");
        for (int v : vet) {
            System.out.print(v + " ");
        }

        System.out.println();

        // Vamos invocar o método passando vetor vet como argumento
        shellSort(vet);
        System.out.println("Apos ShellSort");
        for (int v : vet) {
            System.out.print(v + " ");
        }
    }
}

File: Python/15-Ordenar Vetores/ShellSort.py
#Podemos ordenar um vetor através de alguns métodos criados por programadores
#Este método é chamado ShellSort

def shell_sort(vetor) :
    tam = len(vetor) // 2
    while tam > 0 :
        for i in range(tam) :
            gap_insertion_sort(vetor, i, tam)
        tam = tam // 2
    return vetor

#metodo auxiliar
def gap_insertion_sort(vetor, inicio, gap) :
    for i in range(inicio + gap, len(vetor), gap) :
        valor_atual = vetor[i]
        posicao = i
        while posicao >= gap and vetor[posicao - gap] > valor_atual :
            vetor[posicao] = vetor[posicao - gap]
            posicao -= gap
        vetor[posicao] = valor_atual
    
#Criando vetor
vet = [9,5,7,3,1]
print("Vetor original")

for v in vet :
    print(v, end = " ")

print("\n===========")

#Ordenando com shell sort
print("Shell Sort")
print(shell_sort(vet))
